### DISCUSSION TITLE
Csla 6.2.2 - LazyGetProperty(Async)

### DISCUSSION THREAD

**Discussion by TanguyIngels** [MarimerLLC/csla#3250] [Questions] [answered]
Hi,

Furthermore a request to confirmation.

_First Question_
When you've a parent business object and you want to fetch a property in that BO which is a BusinessBaseList object by a LazyGet, you've to do a Fetch instead of FetchChild?  In point of view you're fetching a child, but I've seen that the DbContext isn't disposed when you do a fetchchild, which can cause dirty data because of the ChangeTracker.   Correct?
But in the Create of the BO you've to do CreateChild? Correct?

```
public static readonly PropertyInfo<Childs> ChildsProperty = RegisterProperty<Childs>(nameof(Childs), RelationshipTypes.LazyLoad);
        public Childs Childs
        {
            get => LazyGetProperty<Childs>(ChildsProperty, () => ApplicationContext.GetRequiredService<IDataPortal<Childs>>().Fetch(Id));
            private set { LoadProperty(ChildsProperty, value); }
        }

[Create]
        [RunLocal]
        private void Create()
        {
            LoadProperty(ChildsProperty, ApplicationContext.GetRequiredService<IChildDataPortal<Childs>>().CreateChild());
            BusinessRules.CheckRules();
        }
```

_Second Question_
When I do LazyGetAsync my ApplicationContext.Identity = null.   Is that normal?   (I'm testing currently in WinForms)

Thanks,

**Comment by rockfordlhotka** [comment, marked as answer]
When lazy loading a child object, you should use a command pattern to invoke the data portal on a root object (the command).

The child data portal does NOT move the call from client to the logical server, so you should NOT be calling the child data portal directly. That causes the problems you describe, plus it will fail if you ever switch to an n-tier deployment.

The command class can look something like this:

```c#
[Serializable]
public class ChildGetter : ReadOnlyBase<ChildGetter>
{
  // declare standard CSLA property for Child - pseudo-code here:
  cslaprop ChildType Child { get; set; }

  [Create]
  private void Create([Inject] IChildDataPortal<Child> dp)
  {
    Child = dp.CreateChild();
  }

  [Fetch]
  private void Fetch(int id, [Inject] IChildDataPortal<Child> dp)
  {
    Child = dp.FetchChild(id);
  }
}
```

Then in your lazy load property implementation, use an `IDataPortal<ChildGetter>` to invoke the command to create or fetch the child.

What happens is that the ChildGetter _does_ move the the logical server, and then it is able to call the _child_ data portal on the logical server, where things are disposed properly, etc. The `Child` property then contains the child object, and when the `ChildGetter` is returned to the logical client, you have access to the lazy-loaded child.
